src/extractor/guidance/intersection_handler.cpp

```diff
@@ -37,13 +38,41 @@ IntersectionHandler::IntersectionHandler(const util::NodeBasedDynamicGraph &node
 {
 }
 
+bool IntersectionHandler::SuppressModeNavigation(const EdgeID &via_edge,
+                                                 const EdgeID &connected,
+                                                 const SuppressModeListT &SUPPRESS_MODE_LIST) const
+{
+    const auto in_mode = node_based_graph.GetEdgeData(via_edge).travel_mode;
+    const auto out_mode = node_based_graph.GetEdgeData(connected).travel_mode;
+    return IntersectionHandler::SuppressModeNavigation(in_mode, out_mode, SUPPRESS_MODE_LIST);
+}
+
+bool IntersectionHandler::SuppressModeNavigation(const TravelMode &in_mode,
+                                                 const TravelMode &out_mode,
+                                                 const SuppressModeListT &SUPPRESS_MODE_LIST) const
+{
+    const auto suppress_in_mode = std::find(begin(SUPPRESS_MODE_LIST), end(SUPPRESS_MODE_LIST), in_mode);
+    if (suppress_in_mode != end(SUPPRESS_MODE_LIST))
+    {
+        return in_mode == out_mode;
+    }
+    return false;
+}
+
 TurnType::Enum IntersectionHandler::findBasicTurnType(const EdgeID via_edge,
                                                       const ConnectedRoad &road) const
 {
 
     const auto &in_data = node_based_graph.GetEdgeData(via_edge);
     const auto &out_data = node_based_graph.GetEdgeData(road.eid);
 
+    const auto in_mode = node_based_graph.GetEdgeData(via_edge).travel_mode;
+    const auto out_mode = node_based_graph.GetEdgeData(road.eid).travel_mode;
+    if (SuppressModeNavigation(in_mode, out_mode, SUPPRESS_MODE_LIST))
```

Comment:
As per https://github.com/Project-OSRM/osrm-backend/pull/3420#issuecomment-265997097, I've added a test. Using the three fallbacks is not sufficient to suppress all turns in ferries. I've added a test that still emits a turn on a ferry.

To fully solve this, I'd propose a dedicated handler (see previous comment). The handler should check all pairs of roads for same travel mode (e.g std::all_of(intersection,SuppressModeNavigation(first_mode,road_mode)) to check if it is applicable and simply assign `NoTurn` to all turns in its `operator()`.